datumaro/plugins/open_images_format.py

```diff
@@ -330,13 +377,114 @@ def _load_bboxes(self, items_by_id):
                         if int_value >= 0:
                             attributes[bool_attr.datumaro_name] = bool(int_value)
 
+                    # Give each box within an item a distinct group ID,
+                    # so that we can later group them together with the corresponding masks.
+                    if (
+                        item.annotations
+                        and item.annotations[-1].type is AnnotationType.bbox
+                    ):
+                        group = item.annotations[-1].group + 1
+                    else:
+                        group = 1
+
                     item.annotations.append(Bbox(
                         label=label_index,
                         x=x_min, y=y_min,
                         w=x_max - x_min, h=y_max - y_min,
                         attributes=attributes,
+                        group=group,
                     ))
 
+                    normalized_coords[id(item.annotations[-1])] \
+                        = np.array([x_min_norm, x_max_norm, y_min_norm, y_max_norm])
+
+        return normalized_coords
+
+    def _load_masks(self, items_by_id, normalized_coords):
+        label_categories = self._categories[AnnotationType.label]
+
+        for mask_path in self._glob_annotations(
+            '*' + OpenImagesPath.MASK_DESCRIPTION_FILE_SUFFIX
+        ):
+            with self._open_csv_annotation(mask_path) as mask_reader:
+                for mask_description in mask_reader:
+                    mask_path = mask_description['MaskPath']
+                    if _RE_INVALID_PATH_COMPONENT.fullmatch(mask_path):
+                        raise UnsupportedMaskPathError(item_id=item.id, mask_path=mask_path)
+
+                    image_id = mask_description['ImageID']
+                    item = items_by_id[image_id]
+
+                    label_name = mask_description['LabelName']
+                    label_index, _ = label_categories.find(label_name)
+                    if label_index is None:
+                        raise UndefinedLabel(
+                            item_id=item.id, subset=item.subset,
+                            label_name=label_name, severity=Severity.error)
+
+                    attributes = {}
+
+                    # The box IDs are rather useless, because the _box_ annotations
+                    # don't include them, so they cannot be used to match masks to boxes.
+                    # However, it is still desirable to record them, because they are
+                    # included in the mask file names, so in order to save each mask to the
+                    # file it was loaded from when saving in-places, we need to know
+                    # the original box ID.
+                    box_id = mask_description['BoxID']
+                    if _RE_INVALID_PATH_COMPONENT.fullmatch(box_id):
+                        raise UnsupportedBoxIdError(item_id=item.id, box_id=box_id)
+                    attributes['box_id'] = box_id
```

Comment:
Maybe these ids should be mask ids?

Reply:
They're called box IDs in the dataset; I don't think we need to rename them. 🤷